### PR TITLE
perlin noise

### PR DESCRIPTION
**Pull request by rafaqz**
**What the pull request does**   
Adds PerlinNoise

**Type of change**   
Please indicate the relevant option(s)

- [ ] :bug: Bug fix (non-breaking change which fixes an issue)
- [x] :sparkle: New feature (non-breaking change which adds functionality)
- [ ] :boom: Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] :book: This change requires a documentation update

**Checklist**
- [x] The changes are documented
  - [x] The docstrings of the different functions describe the arguments, purpose, and behavior 
  - [ ] There are examples in the documentation website
- [ ] The changes are tested
- [x] The changes **do not** modify the behavior of the previously existing functions
  - If they **do**, please explain why and how in the introduction paragraph
- [ ] For **new contributors** - my name and information are added to `.zenodo.json`
- [ ] The `Project.toml` field `version` has been updated
  - Change the *last* number for a `v0.0.x` series release, a bug fix, or a performance improvement
  - Change the *middle* number for additional features that *do not* break the current test suite (unless you fix a bug in the test suite)
  - Change the *first* number for changes that break the current test suite

### CODE REVIEW

**Comment by rafaqz** [comment]
perlin_valley

Perlin looks pretty good

File: src/algorithms/perlinnoise.jl [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+- `periods::Tuple{Int,Int}`: the number of periods of Perlin noise across row and 
+    column dimensions for the first octave.
+- `octaves::Int`: the number of octaves that will form the Perlin noise.
+- `lacunarity::Int` : the rate at which the frequency of periods increases for each octive.
+- `persistance::Real` : The rate at which the amplitude of periods decreases for each octive.
+- `valley::String`: The kind of valley bottom that will be mimicked: `:u` (the defualt) 
+    produces u-shaped valleys, `:v` produces v-shaped valleys, and `:-` 
+    produces flat bottomed valleys.
+"""
+Base.@kwdef struct PerlinNoise{PT<:Real} <: NeutralLandscapeMaker
```
Comment (mkborregaard) [inline src/algorithms/perlinnoise.jl:25]: maybe we should follow the same constructor convention throughout the repo

File: src/algorithms/perlinnoise.jl [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+end
+
+function _landscape!(A, alg::PerlinNoise)
+    # nRow must equal nCol so determine the dimension of the smallest square
+    dim = max(size(A)...)
+    # Check the dim is a multiple of each octives maximum number of periods and 
+    # expand dim if needed
+    rperiodsmax = alg.periods[1] * alg.lacunarity^(alg.octaves - 1)
+    cperiodsmax = alg.periods[2] * alg.lacunarity^(alg.octaves - 1)
+    periodsmultiple = lcm(rperiodsmax, cperiodsmax) # lowest commom multiple
```
Comment (mkborregaard) [inline src/algorithms/perlinnoise.jl]: common not commom

File: src/algorithms/perlinnoise.jl [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    return if alg.valley == :u
+        noise
+    elseif alg.valley == :v
+        abs(noise)
+    elseif alg.valley == :-
+        noise ^ 2
+    end
+end
+
+# This is probably wrong, need to test agains numpy.mgrid
```
Comment (mkborregaard) [inline src/algorithms/perlinnoise.jl]: is this still wrong? Are you concerned about row-major?

File: src/algorithms/perlinnoise.jl [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+meshgrid(x, y) = [i for i in x, j in 1:length(y)], [j for i in 1:length(x), j in y]
+
+function octave_noise(alg::PerlinNoise, octave, nrow, ncol)        
+    f(t) = 6 * t ^ 5 - 15 * t ^ 4 + 10 * t ^ 3 # Wut
```
Comment (mkborregaard) [inline src/algorithms/perlinnoise.jl]: would be good with a more informative comment :joy:

**Comment by mkborregaard** [comment]
looks really cool

File: src/algorithms/perlinnoise.jl (hunk 3 above)
Comment (rafaqz) [reply src/algorithms/perlinnoise.jl]: sorry hadn't pushed the fix yet. It was wrong, but it's right now.

File: src/algorithms/perlinnoise.jl (hunk 4 above)
Comment (rafaqz) [reply src/algorithms/perlinnoise.jl]: thats literally the function from nlmpy, but they have less comments

File: src/algorithms/perlinnoise.jl (hunk 1 above)
Comment (rafaqz) [reply src/algorithms/perlinnoise.jl:25]: Yeah - style in general probably.

**Comment by rafaqz** [comment]
My favourite line so far :

```python
return(octave * (persistence ** octive))
```

`octave` is a matrix, octive is an `Int`. Nice way to grow the namespace

File: src/algorithms/perlinnoise.jl (hunk 4 above)
Comment (mkborregaard) [reply src/algorithms/perlinnoise.jl]: haha

**Comment by rafaqz** [comment]
perlin_delta

File: src/algorithms/perlinnoise.jl (hunk 4 above)
Comment (tpoisot) [reply src/algorithms/perlinnoise.jl]: Still more understandable than [the 2002 original in Java](https://mrl.cs.nyu.edu/~perlin/noise/)

**Comment by tpoisot** [comment]
@rafaqz is this ready to merge?

File: src/algorithms/perlinnoise.jl (hunk 1 above)
Comment (tpoisot) [reply src/algorithms/perlinnoise.jl:25]: Agreed - but maybe @rafaqz approach is more user-friendly?

Comment (rafaqz) [reply src/algorithms/perlinnoise.jl:25]: We could revisit style once we have the bulk of things written and it's clear what the best way is. But I'm happy to change this to whatever is preffered

File: src/algorithms/perlinnoise.jl (hunk 4 above)
Comment (rafaqz) [reply src/algorithms/perlinnoise.jl]: he really went for the magic numbers. But that implementation looks fast, I should have read that first

**Comment by rafaqz** [comment]
Probably is now :)

**Comment by rafaqz** [comment]
~~Hmm actually not sure this is done. The `repeat` based algorithm they use isn't great - it wont handle larger prime numbers without using all my RAM. I think we could write a generator for it - currently it's essentially copying the same data into a huge repeating array for no reason.~~

Edit: actually the basic algorythm does this with primes even without `repeat`. Using repeat just makes it twice as bad (or a bit more than that). 

It probably just needs a warning about using prime numbers for `period`and large array sizes.

**Comment by rafaqz** [comment]
Also whats the plan for tests?

**Comment by mkborregaard** [comment]
> Also whats the plan for tests?

We'll just port the tests from the original package :trollface:

**Review by mkborregaard — APPROVED** [review]
Nice work :+1:

**Comment by mkborregaard** [comment]
Can you rebase on main to resolve the conflict? I'll merge after.

**Review by tpoisot — APPROVED** [review]

**Comment by tpoisot** [comment]
I *think* I have not messed the rebase, and this is ready to merge.

**Comment by rafaqz** [comment]
Added a gallery demo and a few minor tweaks to the methods and docs.

**Comment by rafaqz** [comment]
I think the rebase is borked, I'll fix

**Pull request merged**